enzyme/Enzyme/MLIR/Implementations/MemRefAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,106 @@
+#include "mlir/IR/DialectRegistry.h"
+#include "mlir/Support/LogicalResult.h"
+
+using namespace mlir;
+using namespace mlir::enzyme;
+
+namespace {
+struct LoadOpInterface
+    : public AutoDiffOpInterface::ExternalModel<LoadOpInterface,
+                                                memref::LoadOp> {
+  LogicalResult createForwardModeAdjoint(Operation *op, OpBuilder &builder,
+                                         MGradientUtils *gutils) const {
+    auto loadOp = cast<memref::LoadOp>(op);
+    if (!gutils->isConstantValue(loadOp)) {
+      SmallVector<Value> inds;
+      for (auto ind : loadOp.getIndices())
+        inds.push_back(gutils->getNewFromOriginal(ind));
+      mlir::Value res = builder.create<memref::LoadOp>(
+          loadOp.getLoc(), gutils->invertPointerM(loadOp.getMemref(), builder),
+          inds);
+      gutils->setDiffe(loadOp, res, builder);
+    }
+    gutils->eraseIfUnused(op);
+    return success();
+  }
+};
+
+struct StoreOpInterface
+    : public AutoDiffOpInterface::ExternalModel<StoreOpInterface,
+                                                memref::StoreOp> {
+  LogicalResult createForwardModeAdjoint(Operation *op, OpBuilder &builder,
+                                         MGradientUtils *gutils) const {
+    auto storeOp = cast<memref::StoreOp>(op);
+    if (!gutils->isConstantValue(storeOp.getMemref())) {
+      SmallVector<Value> inds;
+      for (auto ind : storeOp.getIndices())
+        inds.push_back(gutils->getNewFromOriginal(ind));
+      builder.create<memref::StoreOp>(
+          storeOp.getLoc(), gutils->invertPointerM(storeOp.getValue(), builder),
+          gutils->invertPointerM(storeOp.getMemref(), builder), inds);
+    }
+    gutils->eraseIfUnused(op);
+    return success();
+  }
+};
+
+struct AllocOpInterface
+    : public AutoDiffOpInterface::ExternalModel<AllocOpInterface,
+                                                memref::AllocOp> {
+  LogicalResult createForwardModeAdjoint(Operation *op, OpBuilder &builder,
+                                         MGradientUtils *gutils) const {
+    auto allocOp = cast<memref::AllocOp>(op);
+    if (!gutils->isConstantValue(allocOp)) {
+      BlockAndValueMapping map;
+      for (auto op : allocOp->getOperands())
+        map.map(op, gutils->getNewFromOriginal(op));
+      auto nop = builder.clone(*allocOp, map);
+      gutils->setDiffe(allocOp, nop->getResult(0), builder);
+    }
+    gutils->eraseIfUnused(op);
+    return success();
+  }
+};
+
+class MemRefTypeInterface
+    : public AutoDiffTypeInterface::ExternalModel<MemRefTypeInterface,
+                                                  MemRefType> {
+public:
+  Value createNullValue(Type self, OpBuilder &builder, Location loc) const {
+    llvm_unreachable("Cannot create null  of memref (todo polygeist null)");
+  }
+
+  Type getShadowType(Type self, unsigned width) const {
+    assert(width == 1 && "unsupported width != 1");
+    return self;
+  }
+};
+} // namespace
```

Comment:
nit: put the MemRefTypeInterface in a separate file

Reply:
These interfaces aren't externally visible/used, only needed within the implementation (and attached to the Type itself). This is the same layout as the Builtin interface file.